Poly1305 AVX2 backend

str4d opened the pull request:
The AVX2 backend can be selected at compile time with `RUSTFLAGS "-Ctarget-feature=+avx2"`.

str4d commented:
Currently a draft PR because there are still several bugs to fix:
- The "total" test vector from `poly1305-donna` is the only remaining test vector that fail when run on the AVX2 backend, indicating that there is a bug somewhere in the handling of `BlocksX4`.
- I ran AFL to fuzz the AVX2 backend against the software implementation, and it found eight unique differences. I've fixed one of these so far (`x` was not being fully carried in `impl Add<Aligned130> for AdditionKey`).

str4d commented:
Welp, somewhere along the way during my refactoring and documenting process, I've managed to make the AVX2 implementation incredibly slow 😑 When I first got the code running, it was speeding `rage` up from being 2.1x slower than `age` to only 1.3x slower. But now this branch is around 45x slower than the software implementation! I've clearly done something terrible to the underlying AVX2 assembly in my refactor. The main culprit I can think of would be my renaming of variables: the original code reuses variables all over the place, whereas I assigned new variables in many places to clarify the naming.

tarcieri commented:
> The main culprit I can think of would be my renaming of variables: the original code reuses variables all over the place, whereas I assigned new variables in many places to clarify the usage.

Huh, strange, I would assume that LLVM's SSA transform would render these equivalent, at least if the changes were purely superficial...

tarcieri commented:
Regarding runtime selection of backends: we've largely eschewed it so far for a few different reasons.

@newpavlov's suggestion has been to push that selection higher up into e.g. the ChaCha20Poly1305 AEAD construction.

You might check out this (unmerged) PR for runtime detection of PCLMULQDQ for POLYVAL for the past discussion: https://github.com/RustCrypto/universal-hashes/pull/11

str4d commented:
> Huh, strange, I would assume that LLVM's SSA transform would render these equivalent, at least if the changes were purely superficial...

The changes were all just reorganising the existing AVX2 intrinsics to leverage the Rust type system. Next weekend I'll try swapping back in the un-refactored code to confirm it was indeed faster, and then I'll start trying to figure out how the underlying assembly is being altered.

> @newpavlov's suggestion has been to push that selection higher up into e.g. the ChaCha20Poly1305 AEAD construction.

That makes sense to me; we should be consistently selecting the AVX2 backends for the ChaCha20 and Poly1305 components together. I do want auto-detection _somewhere_ though, as it's very inconvenient both as an application developer to be forced to generate separate AVX2 binaries, and as a user to need to care about exactly which link I click to get performance. If someone does want smaller AVX2-only binaries though, it should be possible to obtain them.

str4d commented:
Hmm... it turns out that the AVX2 backend is only ridiculously slow (on my laptop) when compiled without the `+avx2` Rust flag. When I add that, it speeds up to... just slightly slower than the software backend. So there are two issues:

- The runtime detection is not functioning correctly.
  - I'm going to remove it for now, and just rely on the compile-time flag.
- The AVX2 backend is still not faster than the software backend.
  - I'll look at the assembly and try to figure out why. Maybe Rust/LLVM is just really good at vectorising? More likely there are still mistakes in the AVX2 backend.

str4d commented:
I've force-pushed to switch to compile-time backend selection. Unfortunately this means removing the comparative fuzzer; I will rework the crashes it found into regular test cases, so we can still figure them out.

As for the speed issue... it looks to me like the AVX2 backend just generates more assembly than the software backend. It would be useful for someone else to try running this branch to see what their experience is.

tarcieri commented:
> Unfortunately this means removing the comparative fuzzer

FWIW, the `polyval` crate always includes the portable backend even when the CLMUL one is enabled, but sets up type aliases for the former in the event the latter is disabled:

https://github.com/RustCrypto/universal-hashes/blob/master/polyval/src/field.rs

Kinda gross, but I did it for the same reason (at least initially): equivalence testing/fuzzing between the two

str4d commented:
I've rebased this onto current `HEAD`, and... it's now 33% faster than the software backend, with no changes to the AVX2 instructions. No idea what's going on there, but all the tests pass 🤷

str4d commented:
Looking again at the failures I reported earlier:

>     * The "total" test vector from `poly1305-donna` is the only remaining test vector that fail when run on the AVX2 backend, indicating that there is a bug somewhere in the handling of `BlocksX4`.

This test vector was removed by @tarcieri; was it just an invalid vector?
 
>     * I ran AFL to fuzz the AVX2 backend against the software implementation, and it found eight unique differences. I've fixed one of these so far (`x` was not being fully carried in `impl Add<Aligned130> for AdditionKey`).

I re-added the fuzzer helper, and the eight test cases it found previously. Seven of them are still failing.

tarcieri commented:
> This test vector was removed by @tarcieri; was it just an invalid vector?

If there were any I removed which didn't get added back it was a mistake as part of some of the refactoring I was doing (#55)

str4d commented:
I've rebased the PR (so it now contains all the Poly1305 test vectors), and split out my fix for `donna_self_test1` and `fuzz::crash_6` into a separate commit.

I'm working on `fuzz::crash_3` next, but all seven remaining fuzzer crashes exhibit a similar failure mode: if you finalize after some multiple of 4 blocks, the AVX2 backend's accumulator is `0x0100000000` smaller than the software backend's accumulator. My initial guess is that a carry bit from a 32-bit limb is not being correctly propagated somewhere.

str4d commented:
I found the bug! All tests and previous fuzzer crashes now pass 🎉

tarcieri commented on poly1305/src/avx2/helpers.rs:
```diff
@@ -501,7 +501,7 @@ impl Unreduced130 {
                 let t = _mm256_srlv_epi64(v1, _mm256_set_epi64x(64, 64, 64, 26));
                 // v0 + 5·t = [t_3, t_2, t_1, t_0 + 5·(t_4 >> 26)]
-                let red_0 = _mm256_add_epi64(_mm256_add_epi64(v0, t), _mm256_slli_epi32(t, 2));
+                let red_0 = _mm256_add_epi64(_mm256_add_epi64(v0, t), _mm256_slli_epi64(t, 2));
```
Oh my 😅

tarcieri commented:
@str4d want to mark it as ready for review?

str4d commented:
I had a couple of cleanups to do, and I wanted to throw the fuzzer at it again for a few minutes (to shake out any other low-hanging crashes). Done the first, and ~~nothing has com~~ LITERALLY as I was typing this, the fuzzer turned up a crash 😂

str4d commented:
That last crash is now fixed. I found it after 3 minutes of fuzzing, and didn't find anything else after an additional 20 minutes. I'll run some more fuzzing today, but for now I think this is ready to review 😄

tarcieri commented on poly1305/src/avx2.rs:
```diff
@@ -0,0 +1,157 @@
+    r1: PrecomputedMultiplier,
+    r2: PrecomputedMultiplier,
+    initialized: Option<Initialized>,
+    cached_blocks: [u8; BLOCK_X4_SIZE],
```
A bit more of a meta-note, but I'm wondering if the `UniversalHash` trait should have something more like `ParBlocks` on `UniversalHash` which can be used to signal to the consumer of the trait how many blocks the underlying implementation can work on at a time:

https://docs.rs/block-cipher/0.8.0/block_cipher/trait.BlockCipher.html#associatedtype.ParBlocks

tarcieri replied:
Opened an issue about this for the upstream `universal-hash` crate: https://github.com/RustCrypto/traits/issues/289

Let's continue discussion there. Otherwise I'd say this is fine for now and we should figure out a general solution upstream.

tarcieri commented on poly1305/src/avx2.rs:
```diff
@@ -0,0 +1,157 @@
+            return;
+        }
+
+        self.cached_blocks
+            [self.num_cached_blocks * BLOCK_SIZE..(self.num_cached_blocks + 1) * BLOCK_SIZE]
+            .copy_from_slice(block);
```
This seems like the sort of thing that might benefit from the cache being a `[Block; 4]`, although it seems it might complicate some of the cases below. Six of one, half dozen of the other.

str4d replied:
I went for a single `[u8; 4 * BLOCK_SIZE]` because `_mm256_loadu_si256` (inside `Aligned4x130::from_blocks`) takes a pointer to a contiguous 32-byte array, and I wasn't sure whether `[Block; 4]` would be contiguous.

tarcieri replied:
To my knowledge, Rust array layout is fully guaranteed in these cases (i.e. size = stride), even with `generic-array` (which is just a wrapper for the underlying array types which uses type-level operations to choose which size to use).

I believe safe conversions for things like `&[[T; N]] -> &[T]` are one of the goals of the [safe transmute project](https://internals.rust-lang.org/t/pre-rfc-v2-safe-transmute/11431).

tarcieri replied:
More info here: https://github.com/rust-lang/unsafe-code-guidelines/blob/1c44b54/reference/src/layout/arrays-and-slices.md

tarcieri replied:
I'll go ahead and merge. I can take a look at following up on this before cutting a release unless you'd like to.

tarcieri commented on poly1305/src/fuzz.rs:
```diff
@@ -0,0 +1,153 @@
+use super::fuzz_avx2;
```
Re: this module, just curious: what fuzzer were you actually using to find these failing cases? Should we add a toplevel `fuzz/` directory?

str4d replied:
I used [afl.rs](https://github.com/rust-fuzz/afl.rs), which appears to prefer that its fuzzing crate be separate from the main crate (or at least, that's how I interpreted its docs).

tarcieri replied:
Do these changes just magically work with afl.rs somehow, or is there an associated config needed to run the fuzzer?

str4d replied:
[The tutorial](https://rust-fuzz.github.io/book/afl/tutorial.html) covers most of it, but you do need to specify a `main.rs`. I used this:
```
#[macro_use]
extern crate afl;

fn main() {
    fuzz!(|data: &[u8]| {
        // Use first 32 bytes of data as key.
        if data.len() >= 32 {
            poly1305::fuzz_avx2((&data[0..32]).into(), &data[32..]);
        }
    });
}
```
I could inline the length-check and data-splitting into `poly1305::fuzz_avx2` so that the helper here is the most obvious possible thing?

tarcieri replied:
Could you throw that code into something like `fuzz/main.rs`?
